File: services/anomaly-detector/catalog.go
package main

import (
	"context"
	"fmt"
	"sort"
	"sync"

	"github.com/jackc/pgx/v5/pgxpool"
)

type metricRange struct {
	Min, Max float64
}

type deviceInfo struct {
	MachineType string
	Ranges      map[string]metricRange // metric -> operating range, for rule-based detection
}

// catalog holds a periodically-refreshed snapshot of device metadata needed
// for detection: each device's machine type (for grouping isolation forests)
// and each sensor's operating range (for rule-based thresholds). It's kept
// in memory and refreshed on a timer rather than queried per-event, since
// this metadata changes rarely relative to telemetry volume.
type catalog struct {
	mu           sync.RWMutex
	devices      map[string]deviceInfo // device_id -> info
	featureOrder map[string][]string   // machine_type -> ordered metric names (the isolation forest's feature vector shape)
	pool         *pgxpool.Pool
}

func newCatalog(ctx context.Context, dsn string) (*catalog, error) {
	pool, err := pgxpool.New(ctx, dsn)
	if err != nil {
		return nil, fmt.Errorf("connect to postgres: %w", err)
	}
	c := &catalog{
		devices:      make(map[string]deviceInfo),
		featureOrder: make(map[string][]string),
		pool:         pool,
	}
	if err := c.refresh(ctx); err != nil {
		pool.Close()
		return nil, err
	}
	return c, nil
}

func (c *catalog) close() {
	c.pool.Close()
}

func (c *catalog) refresh(ctx context.Context) error {
	rows, err := c.pool.Query(ctx, `
		SELECT d.id, m.machine_type, s.metric,
		       COALESCE(s.min_operating_value, 0), COALESCE(s.max_operating_value, 100)
		FROM devices d
		JOIN machines m ON m.id = d.machine_id
		JOIN sensors s ON s.device_id = d.id
	`)
	if err != nil {
		return fmt.Errorf("query catalog: %w", err)
	}
	defer rows.Close()

	devices := make(map[string]deviceInfo)
	metricsByType := make(map[string]map[string]bool)

	for rows.Next() {
		var deviceID, machineType, metric string
		var min, max float64
		if err := rows.Scan(&deviceID, &machineType, &metric, &min, &max); err != nil {
			return fmt.Errorf("scan catalog row: %w", err)
		}

		info, ok := devices[deviceID]
		if !ok {
			info = deviceInfo{MachineType: machineType, Ranges: make(map[string]metricRange)}
		}
		info.Ranges[metric] = metricRange{Min: min, Max: max}
		devices[deviceID] = info

		if metricsByType[machineType] == nil {
			metricsByType[machineType] = make(map[string]bool)
		}
		metricsByType[machineType][metric] = true
	}
	if err := rows.Err(); err != nil {
		return fmt.Errorf("iterate catalog rows: %w", err)
	}

	featureOrder := make(map[string][]string, len(metricsByType))
	for machineType, metrics := range metricsByType {
		ordered := make([]string, 0, len(metrics))
		for m := range metrics {
			ordered = append(ordered, m)
		}
		sort.Strings(ordered) // deterministic feature-vector dimension order
		featureOrder[machineType] = ordered
	}

	c.mu.Lock()
	c.devices = devices
	c.featureOrder = featureOrder
	c.mu.Unlock()
	return nil
}

func (c *catalog) lookup(deviceID string) (deviceInfo, bool) {
	c.mu.RLock()
	defer c.mu.RUnlock()
	info, ok := c.devices[deviceID]
	return info, ok
}

func (c *catalog) featuresFor(machineType string) []string {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.featureOrder[machineType]
}

File: services/anomaly-detector/catalog.py
"""Device/sensor metadata catalog, the Python port of catalog.go.

Holds a periodically-refreshed snapshot of device metadata needed for
detection: each device's machine type (for grouping isolation forests) and
each sensor's operating range (for rule-based thresholds). It's kept in
memory and refreshed on a timer rather than queried per-event, since this
metadata changes rarely relative to telemetry volume.
"""

from __future__ import annotations

import threading

from psycopg_pool import ConnectionPool

from rules import MetricRange


class DeviceInfo:
    def __init__(self, machine_type: str):
        self.machine_type = machine_type
        self.ranges: dict[str, MetricRange] = {}


_CATALOG_QUERY = """
    SELECT d.id, m.machine_type, s.metric,
           COALESCE(s.min_operating_value, 0), COALESCE(s.max_operating_value, 100)
    FROM devices d
    JOIN machines m ON m.id = d.machine_id
    JOIN sensors s ON s.device_id = d.id
"""


class Catalog:
    """Opens this service's only Postgres connection pool, sized explicitly
    via max_conns rather than a library default (see
    ANOMALY_POSTGRES_MAX_CONNS in .env.example). Raises if the initial
    refresh fails, matching Go's newCatalog fail-fast startup behavior —
    the caller is expected to treat that as fatal."""

    def __init__(self, dsn: str, max_conns: int):
        self._lock = threading.Lock()
        self._devices: dict[str, DeviceInfo] = {}
        self._feature_order: dict[str, list[str]] = {}
        self._pool = ConnectionPool(dsn, min_size=1, max_size=max_conns, open=True)
        self.refresh()

    def close(self) -> None:
        self._pool.close()

    def pool(self) -> ConnectionPool:
        """Exposes the underlying pool for idempotency.py's claim query,
        which needs a raw connection outside the catalog's own read/refresh
        paths — mirrors Go's cat.pool field access from main.go (same
        package there; a small accessor here, since Python module
        boundaries make a bare attribute reach-through less idiomatic)."""
        return self._pool

    def ping(self) -> bool:
        try:
            with self._pool.connection(timeout=2.0) as conn:
                conn.execute("SELECT 1")
            return True
        except Exception:  # noqa: BLE001
            return False

    def refresh(self) -> None:
        with self._pool.connection() as conn:
            rows = conn.execute(_CATALOG_QUERY).fetchall()

        devices: dict[str, DeviceInfo] = {}
        metrics_by_type: dict[str, set[str]] = {}

        for device_id, machine_type, metric, lo, hi in rows:
            # psycopg3 returns Postgres `uuid` columns as uuid.UUID objects,
            # not str — cast explicitly so lookup() (keyed by the plain
            # device_id string every event carries) actually matches.
            device_id = str(device_id)
            info = devices.get(device_id)
            if info is None:
                info = DeviceInfo(machine_type)
                devices[device_id] = info
            info.ranges[metric] = MetricRange(min=float(lo), max=float(hi))
            metrics_by_type.setdefault(machine_type, set()).add(metric)

        feature_order = {mt: sorted(metrics) for mt, metrics in metrics_by_type.items()}

        with self._lock:
            self._devices = devices
            self._feature_order = feature_order

    def lookup(self, device_id: str) -> DeviceInfo | None:
        with self._lock:
            return self._devices.get(device_id)

    def features_for(self, machine_type: str) -> list[str]:
        with self._lock:
            return self._feature_order.get(machine_type, [])
